<div id="container" style="position:relative;">
<div style="float:centre"><h1> Yelp Review Analysis </h1></div>
<div style="position:relative; float:right"><img style="height:65px" src ="https://drive.google.com/uc?export=view&id=1EnB0x-fdqMp6I5iMoEBBEuxB_s7AmE2k" />
</div>
<div style="position:relative; float:left"><h4> by Lavanya Kwatra (August, 2020) </h4></div>
</div>

## Introduction:

Founded in October 2004, Yelp is an online platform that publishes crowd-sourced reviews for businesses, along with the recent addition of Yelp Reservations, which is an online booking service. Headquartered in San Francisco, California, Yelp develops, hosts and markets their website (www.yelp.com) as well as the Yelp mobile app. Yelp prides itself on their motto of “connecting people with great local businesses,” thereby, creating a marketplace within their user community to study the social perception of the businesses they advertise. (Yelp, Inc., n.d.)

As a result of this marketplace model, in a way, businesses advertise to new customers through the user-generated reviews on Yelp and rely heavily on consumer engagement, such as “useful votes” on the review another user wrote about their establishment. Yelp generates insights from their online community of users, who write reviews & rate businesses on a scale of 1 to 5 while interacting with other users & the posted content in a social networking format. 

A traditional consumer centric approach would focus on tracking user activity levels, such as:
1.	the total number of reviews written by the user,
2.	whether the user leaves "tips" or suggestions for the restaurant online,
3.	whether the user checks into restaurants when they dine there,
4.	the amount of time the user spends on the platform, 
5.	the user’s social interaction with other users through “votes” on reviews
6.	whether the user actively finds friends in the user community,
7.	whether the user follows top Yelper's or similar users,


The most popular reason people use Yelp is when they are searching for a new restaurant, based on how other users found it. This is captured in the user “votes” on reviews posted by other users in the Yelp community. Based on a [study](https://kanuparthy.files.wordpress.com/2016/05/cscwf630.pdf) by Saeideh Bakhshi, Partha Kanuparthy and David A. Shamma at Yahoo Labs, a deeper understanding of the social perception of these "votes" as user feedback can help design better recommendation and social networks. Using the same review data from 2013, they found that active and older members tend to engage more on the platform, by writing more reviews and reacting to other reviews. Reviews with a higher word count are perceived as more "useful" overall by the community, while those who give higher ratings are perceived as "cool" and review with a negative tone are more likely to be voted "funny" by other users. Additionally, they found that new businesses can generate a "review momentum" by receiving more "useful" votes on initial reviews by users, which helps establish the business. (Bakhshi, Kanuparthy, & Shamma, 2013)

------

# Data Selection

Yelp’s website has a well-documented dataset of 6 .json files, namely: business, user, review, tips, check-ins and photos. 

Based on which data is relevant to study the social perception of restaurants in the Yelp user community, `review.json`, `business.json` and `user.json` were downloaded to the local drive. The three files are almost 11GB in size, containing over 8 million reviews for 200,000+ businesses across North America. 

Note: these cells below take a while to run and will require downloading the .json files that are hosted in an S3 bucket, `yelp` on AWS. 



In [1]:
# reading in the datasets 
# downloaded from Yelp as .json files
# importing pandas to read the files into a dataframe
import pandas as pd

In [2]:
# reading in business attributes data
business = pd.read_json(r'business.json', lines=True)

In [3]:
# reading in user attributes data
user = pd.read_json(r'user.json', lines=True)

In [4]:
# reading in review attributes data
review = pd.read_json(r'review.json', lines=True)

In [5]:
# viewing the head of the user dataframe
user.head()

user_id      name  review_count        yelping_since  \
0  ntlvfPzc8eglqvk92iDIAw    Rafael           553  2007-07-06 03:27:11   
1  FOBRPlBHa3WPHFB5qYDlVg  Michelle           564  2008-04-28 01:29:25   
2  zZUnPeh2hEp0WydbAZEOOg    Martin            60  2008-08-28 23:40:05   
3  QaELAmRcDc5TfJEylaaP8g      John           206  2008-09-20 00:08:14   
4  xvu8G900tezTzbbfqmTKvA      Anne           485  2008-08-09 00:30:27   

   useful  funny  cool                                         elite  \
0     628    225   227                                                 
1     790    316   400                 2008,2009,2010,2011,2012,2013   
2     151    125   103                                          2010   
3     233    160    84                                          2009   
4    1265    400   512  2009,2010,2011,2012,2014,2015,2016,2017,2018   

                                             friends  fans  ...  \
0  oeMvJh94PiGQnx_6GlndPQ, wm1z1PaJKvHgSDRKfwhfDg...    14  ...   
1  ly7EnE8leJmyqyePVYFlug, pRlR63iDytsnnniPb3AOug...    27  ...   
2  Uwlk0txjQBPw_JhHsQnyeg, Ybxr1tSCkv3lYA0I1qmnPQ...     5  ...   
3  iog3Nyg1i4jeumiTVG_BSA, M92xWY2Vr9w0xoH8bPplfQ...     6  ...   
4  3W3ZMSthojCUirKEqAwGNw, eTIbuu23j9tOgmIa9POyLQ...    78  ...   

   compliment_more  compliment_profile  compliment_cute  compliment_list  \
0                2                   1                0                1   
1                4                   5                2                1   
2                6                   0                1                0   
3                1                   0                0                0   
4                9                   2                1                1   

   compliment_note  compliment_plain  compliment_cool  compliment_funny  \
0               11                15               22                22   
1               33                37               63                63   
2                3                 7               17                17   
3                7                14                7                 7   
4               22                28               31                31   

   compliment_writer  compliment_photos  
0                 10                  0  
1                 21                  5  
2                  4                  1  
3                  2                  0  
4                 19                 31  

[5 rows x 22 columns]

In [6]:
# viewing the head of the business dataframe
business.head()

business_id                      name  \
0  f9NumwFMBDn751xgFiRbNA  The Range At Lake Norman   
1  Yzvjg0SayhoZgCljUJRF9Q         Carlos Santo, NMD   
2  XNoUzKckATkOD1hP6vghZg                   Felinus   
3  6OAZjbxqM5ol29BuHsil3w      Nevada House of Hose   
4  51M2Kk903DFYI6gnB5I6SQ   USE MY GUY SERVICES LLC   

                     address             city state postal_code   latitude  \
0            10913 Bailey Rd        Cornelius    NC       28031  35.462724   
1  8880 E Via Linda, Ste 107       Scottsdale    AZ       85258  33.569404   
2      3554 Rue Notre-Dame O         Montreal    QC     H4C 1P4  45.479984   
3             1015 Sharp Cir  North Las Vegas    NV       89030  36.219728   
4         4827 E Downing Cir             Mesa    AZ       85205  33.428065   

    longitude  stars  review_count  is_open  \
0  -80.852612    3.5            36        1   
1 -111.890264    5.0             4        1   
2  -73.580070    5.0             5        1   
3 -115.127725    2.5             3        0   
4 -111.726648    4.5            26        1   

                                          attributes  \
0  {'BusinessAcceptsCreditCards': 'True', 'BikePa...   
1  {'GoodForKids': 'True', 'ByAppointmentOnly': '...   
2                                               None   
3  {'BusinessAcceptsCreditCards': 'True', 'ByAppo...   
4  {'BusinessAcceptsCreditCards': 'True', 'ByAppo...   

                                          categories  \
0  Active Life, Gun/Rifle Ranges, Guns & Ammo, Sh...   
1  Health & Medical, Fitness & Instruction, Yoga,...   
2                   Pets, Pet Services, Pet Groomers   
3  Hardware Stores, Home Services, Building Suppl...   
4  Home Services, Plumbing, Electricians, Handyma...   

                                               hours  
0  {'Monday': '10:0-18:0', 'Tuesday': '11:0-20:0'...  
1                                               None  
2                                               None  
3  {'Monday': '7:0-16:0', 'Tuesday': '7:0-16:0', ...  
4  {'Monday': '0:0-0:0', 'Tuesday': '9:0-16:0', '...

In [7]:
# viewing the head of the review dataframe
review.head()

review_id                 user_id             business_id  \
0  xQY8N_XvtGbearJ5X4QryQ  OwjRMXRC0KyPrIlcjaXeFQ  -MhfebM0QIsKt87iDN-FNw   
1  UmFMZ8PyXZTY2QcwzsfQYA  nIJD_7ZXHq-FX8byPMOkMQ  lbrU8StCq3yDfr-QMnGrmQ   
2  LG2ZaYiOgpr2DK_90pYjNw  V34qejxNsCbcgD8C0HVk-Q  HQl28KMwrEKHqhFrrDqVNQ   
3  i6g_oA9Yf9Y31qt0wibXpw  ofKDkJKXSKZXu5xJNGiiBQ  5JxlZaqCnk1MnbgRirs40Q   
4  6TdNDKywdbjoTkizeMce8A  UgMW8bLE0QMJDCkQ1Ax5Mg  IS4cv902ykd8wj1TR0N3-A   

   stars  useful  funny  cool  \
0      2       5      0     0   
1      1       1      1     0   
2      5       1      0     0   
3      1       0      0     0   
4      4       0      0     0   

                                                text                date  
0  As someone who has worked with many museums, I... 2015-04-15 05:21:16  
1  I am actually horrified this place is still in... 2013-12-07 03:16:52  
2  I love Deagan's. I do. I really do. The atmosp... 2015-12-05 03:18:11  
3  Dismal, lukewarm, defrosted-tasting "TexMex" g... 2011-05-27 05:30:52  
4  Oh happy day, finally have a Canes near my cas... 2017-01-14 21:56:57

In [ ]:
# Lets view the shapes of each of these dataframes:

In [8]:
business.shape

(209393, 14)

In [9]:
user.shape

(1968703, 22)

In [10]:
review.shape

(8021122, 9)

From the above, the following conclusions are drawn:

- We have 8,021,122 reviews with 9 corresponding review features, 209,393 businesses with 14 corresponding business attributes and a user base of 1,968,703 users with 22 corresponding user dimensions.

- For each review, there is a primary key called `review_id`, along with the corresponding `business_id`, relating it to the business and `user_id`, relating it to the user who wrote the review about that business.

- In order to create a dataset that would be ideal for the model to classify the sentiments based on a cluster analysis of users, it is important to merge these three files to concatenate all the attributes pertaining to that review in one row of the dataframe.

In [11]:
# merging reviews with business on business_id 
# Note: set index as the business_id to avoid adding the column again
review = review.merge(business.set_index('business_id'), on='business_id')

In [12]:
# merging the new review with business attributes with user on user_id
# Note: set index as the user_id to avoid adding the column again
review = review.merge(user.set_index('user_id'), on='user_id')

In [13]:
# viewing the shape of the new complete review dataframe 
# with the corresponding business & user attributes
# Note that the number of reviews is the same, therefore, same # of rows
# All the columns from the two dataframes have been added to review
review.shape

(8021122, 43)

In [14]:
# to view all column names in the head
pd.set_option('display.max_columns', None)

In [15]:
# viewing head of the dataframe for sanity check
review.head()

review_id                 user_id             business_id  \
0  xQY8N_XvtGbearJ5X4QryQ  OwjRMXRC0KyPrIlcjaXeFQ  -MhfebM0QIsKt87iDN-FNw   
1  SjfnCrMCgOiWafnQuCKlhw  OwjRMXRC0KyPrIlcjaXeFQ  9SU7ZZhaFUJJ6m2k5HKHeg   
2  t7xOZF5UKXjSpVcXLOSAgw  owbC7FP8SNAlwv6f9S5Stw  -MhfebM0QIsKt87iDN-FNw   
3  WE2gu5ww_FOR8IYWb10GFQ  owbC7FP8SNAlwv6f9S5Stw  4ZbRwCB9oGibxK21MUZKHA   
4  Djql0bAS55_sl7im29wD0g  owbC7FP8SNAlwv6f9S5Stw  4ZbRwCB9oGibxK21MUZKHA   

   stars_x  useful_x  funny_x  cool_x  \
0        2         5        0       0   
1        1         4        0       0   
2        2         2        2       0   
3        2         1        3       0   
4        1         1        0       0   

                                                text                date  \
0  As someone who has worked with many museums, I... 2015-04-15 05:21:16   
1  SLS just opened in August and they have so man... 2015-03-19 06:17:28   
2  I have been there. I believe more than once. \... 2014-03-14 08:24:25   
3  This Smiths sucks ass.\nThey invented Yelp for... 2015-06-23 06:30:12   
4  I have an update ! \nDo not EVER pay your util... 2018-02-10 21:51:52   

                         name_x                address       city state  \
0  Bellagio Gallery of Fine Art  3600 S Las Vegas Blvd  Las Vegas    NV   
1                        Sahara  2535 Las Vegas Blvd S  Las Vegas    NV   
2  Bellagio Gallery of Fine Art  3600 S Las Vegas Blvd  Las Vegas    NV   
3                       Smith's   9750 S Maryland Pkwy  Las Vegas    NV   
4                       Smith's   9750 S Maryland Pkwy  Las Vegas    NV   

  postal_code   latitude   longitude  stars_y  review_count_x  is_open  \
0       89109  36.112896 -115.177637      3.5             180        1   
1       89109  36.142375 -115.156723      3.0            2259        1   
2       89109  36.112896 -115.177637      3.5             180        1   
3       89183  36.012129 -115.134257      2.5              96        1   
4       89183  36.012129 -115.134257      2.5              96        1   

                                          attributes  \
0  {'BusinessAcceptsCreditCards': 'True', 'Restau...   
1  {'BikeParking': 'False', 'Ambience': '{'romant...   
2  {'BusinessAcceptsCreditCards': 'True', 'Restau...   
3  {'RestaurantsTakeOut': 'True', 'BikeParking': ...   
4  {'RestaurantsTakeOut': 'True', 'BikeParking': ...   

                                          categories  \
0  Shopping, Arts & Entertainment, Art Galleries,...   
1  Hotels & Travel, Nightlife, Hotels, Event Plan...   
2  Shopping, Arts & Entertainment, Art Galleries,...   
3  Beauty & Spas, Delis, Grocery, Cosmetics & Bea...   
4  Beauty & Spas, Delis, Grocery, Cosmetics & Bea...   

                                               hours name_y  review_count_y  \
0  {'Monday': '10:0-20:0', 'Tuesday': '10:0-20:0'...  Jamie              58   
1  {'Monday': '0:0-0:0', 'Tuesday': '0:0-0:0', 'W...  Jamie              58   
2  {'Monday': '10:0-20:0', 'Tuesday': '10:0-20:0'...   Girl             250   
3  {'Monday': '6:0-1:0', 'Tuesday': '6:0-1:0', 'W...   Girl             250   
4  {'Monday': '6:0-1:0', 'Tuesday': '6:0-1:0', 'W...   Girl             250   

         yelping_since  useful_y  funny_y  cool_y elite  \
0  2008-12-18 09:41:38        85       21      28         
1  2008-12-18 09:41:38        85       21      28         
2  2012-09-28 19:36:16       507      265     185         
3  2012-09-28 19:36:16       507      265     185         
4  2012-09-28 19:36:16       507      265     185         

                                             friends  fans  average_stars  \
0  m2R7pydiZ7vezHD-oZt6-g, EpAkqFJLjztmlAscrTfh_Q...     4           3.36   
1  m2R7pydiZ7vezHD-oZt6-g, EpAkqFJLjztmlAscrTfh_Q...     4           3.36   
2  3tLPT5PzOemSed-ERxwSQg, TuivYYtdK6y2uOgbE2axXA...     8           3.42   
3  3tLPT5PzOemSed-ERxwSQg, TuivYYtdK6y2uOgbE2axXA...     8           3.42   
4  3tLPT5PzOemSed-ERxwSQg, TuivYYtdK6y2uOgbE2axXA...     8      

In [16]:
# taking a sample of this dataframe
# random_state set to get the same points & take a representative sample
sample = review.sample(frac=0.0625, random_state=11)

In [7]:
# exporting this sample as a .csv file to local drive
sample.to_csv('sample.csv')

Now that the reviews have been matched to their corresponding business & user attributes, we can begin to explore the data to understand the social perception of users in the Yelp community with respect to restaurants. The next notebook will deal with further cleaning of this data.

------------------------------------------------